src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1846,6 +1846,51 @@ void MergeTreeData::clearPartsFromFilesystemImpl(const DataPartsVector & parts_t
     }
 }
 
+size_t MergeTreeData::clearOldBrokenPartsFromDetachedDirecory()
```

Comment:
A long comment that proves that it's safe is required. It must contain a note about mtime and renames, explain why it's not a problem, mention scenarios when it might be a problem and why we don't care about these scenarios.